cbits/reverse.c

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2021 Andrew Lelechenko <andrew.lelechenko@gmail.com>
+ */
+
+#include <string.h>
+#include <stdint.h>
+
+void _hs_text_reverse(uint8_t *dst0, const uint8_t *src0, size_t off, size_t len)
+{
+  const uint8_t *src = src0 + off;
+  const uint8_t *srcend = src + len;
+  uint8_t *dst = dst0 + len - 1;
+
+  while (src < srcend){
```

Comment:
Yes. Thanks, I've added a comment.

Reply:
Thanks for the comments. I think it may be useful to add a similar comment for the FFI wrappers, just in case.  I gather that it is safe to assume that in any `Text arr off len` structure the library strictly ensures that the `let` bytes starting at `off` in `are` are a valid UTF-8 sequence?

Perhaps consumers of the `Data.Text.Internal` API could benefit from a warning to maintain the requisite invariant, since neither the exported raw `Text` constructor, nor the `text_` _smart constructor_ enforce this property.
```haskell
-- | A space efficient, packed, unboxed Unicode text type.
data Text = Text
    {-# UNPACK #-} !A.Array          -- bytearray encoded as UTF-8
    {-# UNPACK #-} !Int              -- offset in bytes (not in Char!)
    {-# UNPACK #-} !Int              -- length in bytes (not in Char!)
    deriving (Typeable, TH.Lift)

-- | Smart constructor.
text_ ::
#if defined(ASSERTS)
  HasCallStack =>
#endif
  A.Array -> Int -> Int -> Text
text_ arr off len =
#if defined(ASSERTS)
  let c    = A.unsafeIndex arr off
  in assert (len >= 0) .
     assert (off >= 0) .
     assert (len == 0 || c < 0x80 || c >= 0xC0) $
#endif
     Text arr off len
{-# INLINE text_ #-}

-- | Construct a 'Text' without invisibly pinning its byte array in
-- memory if its length has dwindled to zero.
text ::
#if defined(ASSERTS)
  HasCallStack =>
#endif
  A.Array -> Int -> Int -> Text
text arr off len | len == 0  = empty
                 | otherwise = text_ arr off len
{-# INLINE text #-}

textP :: A.Array -> Int -> Int -> Text
{-# DEPRECATED textP "Use text instead" #-}
textP = text

-- $internals
-- 
-- Internally, the 'Text' type is represented as an array of 'Word8'
-- UTF-8 code units. The offset and length fields in the constructor
-- are in these units, /not/ units of 'Char'.
--
-- Invariants that all functions must maintain:
-- 
-- * Since the 'Text' type uses UTF-8 internally, it cannot represent
--   characters in the reserved surrogate code point range U+D800 to
--   U+DFFF. To maintain this invariant, the 'safe' function maps
--   'Char' values in this range to the replacement character (U+FFFD,
--   \'&#xfffd;\').
```

The comment in `$internals` could perhaps forcefully explain that wrapping byte array slices that are not valid UTF-8 data as `Text` objects leads to **undefined** behaviour.

Reply:
Valid point, thanks! I'll update soon.